### ISSUE TITLE
ActiveRecord calling save on a parent with accepts_nested_attributes_for with before_destroy on child model can raise error

### ISSUE THREAD

**Issue by rschooley** [rails/rails#31431]
### Steps to reproduce
have a parent model with `accepts_nested_attributes_for :foos, allow_destroy: true`
have the child model add an error and return false in a `before_destroy` hook
mark the child with `_destroy` in a `fields_for`
save the form

### Expected behavior
Calling `save` on the parent should return false.

### Actual behavior
An unhandled exception is thrown `ActiveRecord::RecordNotDestroyed`

### System configuration
Rails version: 4.2.10
Ruby version: 2.4.1

Here is a link to sample project reproducing it, with instructions in the Readme.  Let me know if I can add any more useful info, thanks.
https://github.com/rschooley/rails_nested_sample

**Comment by y-yagi** [comment]
Please check if the issue is reproduced with latest Rails. 
Rails 4.2.x only supports severe security issues. 
If you can be reproduced the issue with latest Rails, feel free to reopen this issue. Thanks!

**Comment by rschooley** [comment]
This has been an issue going back over at least a year (rails 4 to rails 5) and should not be closed until a PR fixes it.

**Comment by rafaelfranca** [comment]
@rschooley

**Comment by why-el** [comment]
@rafaelfranca I don't think this is a bug. The `before_destroy` on the associated task model is using `throw(:abort)` (which is not caught anywhere), so the model cannot be saved, rightly so.

**Comment by filipegiusti** [comment]
https://github.com/rails/rails/commit/5aab0c053832ded70a3a4b58cb97f8f8bba796ba seems to be when the bug was introduced and it's still [a thing](https://github.com/rails/rails/blob/f41730c95ac828ff228291df188b04a3ca7c3787/activerecord/lib/active_record/associations/has_many_association.rb#L99) in latest Rails.

Apparently [this](https://github.com/rails/rails/blob/3dff30c14fa25f211088d62da26611616ab5616f/activerecord/lib/active_record/associations/collection_association.rb#L192-L199) needs to differentiate between `destroy` and `destroy!` and might be related to https://github.com/rails/rails/issues/35298.
